extensions/amp-viewer-integration/0.1/findtext.js

```diff
@@ -0,0 +1,451 @@
+  if (start > 0) {
+    parent.insertBefore(document.createTextNode(
+        text.substring(0, start)), node);
+  }
+  const span = document.createElement('span');
+  span.appendChild(document.createTextNode(text.substring(start, end)));
+  parent.insertBefore(span, node);
+  marked.push(span);
+
+  let endText = null;
+  if (end < text.length) {
+    endText = document.createTextNode(text.substring(end));
+    parent.insertBefore(endText, node);
+  }
+  parent.removeChild(node);
+  return endText;
+};
+
+
+/**
+ * nextTextNode finds the next sibling text node or
+ *   the next text node in the siblings of the parent.
+ * @param {!Text} textNode The node to start to find the next text node.
+ * @return {?Text}
+ */
+const nextTextNode = function(textNode) {
+  // If leaving is true, find the next node from siblings or the parent.
+  // If leaving is false, find the next node from childrens.
+  let leaving = true;
+  /** @type {Node} */
+  let node = textNode;
+  while (true) {
+    if (node == null) {
+      // No more parent.
+      return null;
+    }
+    if (leaving) {
+      const next = node.nextSibling;
+      if (next) {
+        node = next;
+        // visits childrens of node.
+        leaving = false;
+      } else {
+        // back to the parent.
+        node = node.parentNode;
+      }
+      continue;
+    }
+    if (node instanceof Text) {
+      return node;
+    }
+    if (!node.firstChild) {
+      leaving = true;
+    } else {
+      node = node.firstChild;
+    }
+  }
+};
+
+
+/**
+ * A special TextPos object to represent a whitespace injected
+ *   between two block nodes.
+ * @type {!TextPos}
+ */
+const posDomDelimiter = new TextPos(document.createTextNode(' '), 0);
+
+/**
+ * TextScanner visits text nodes under a root node and
+ *   returns charcter positions respectively.
+ */
+export class TextScanner {
+  /**
+   * @param {!Node} root The root node to visit.
+   */
+  constructor(root) {
+    this.internal_ = new TextScannerInternal(root, false);
+    this.next_ = this.internal_.next();
+  }
+
+  /**
+   * next returns the next TextPos.
+   * Returns null when the scanner reaches the end of the text.
+   * @return {?TextPos}
+   */
+  next() {
+    const ret = this.next_;
+    if (ret == null) {
+      return null;
+    }
+    this.next_ = this.internal_.next();
+    if (this.next_ == null && /\s/.test(ret.char)) {
+      // Remove the trailing space.
+      return null;
+    }
+    return ret;
+  }
+}
+
+export class TextScannerInternal {
+  /**
+   * @param {!Node} node The root node to visit.
+   * @param {boolean} needSpace Whether spaces should be output if found.
+   */
+  constructor(node, needSpace) {
+    this.node_ = node;
+    this.needSpace_ = needSpace;
+
+    this.textIdx_ = -1;
+    this.child_ = null;
+    this.putDomDelim_ = false;
+    if (node instanceof Text) {
+      this.textIdx_ = 0;
+    } else if (node instanceof Element) {
+      const style = computedStyle(window, node);
```

Comment:
This forces repaint. I don't think it's necessary or accurate to check if the element is `display: none`, as the element could be hidden by other means (opacity, sizing, etc).

Reply:
> This forces repaint.

Do you mean reflow (layout)?

According to [this doc](https://www.google.com/search?q=force+reflow&hl=en), getComputedStyle forces relayout only when the CSS properties that require layouts are accessed unless media queries are used.

I created a simple example to check getComputedStyle(node).display (https://jsfiddle.net/7c7rq2ot/). You can see the reading `display` is faster than `width` because browsers do not relayout elements when `display` is read. I checked it with iPhone Safari, Android Chrome.

I confirmed reading `display` does not force layout with Performance analyzer of Google Chrome devtools.
<img width="540" alt="nolayout" src="https://user-images.githubusercontent.com/255937/40767653-19f44934-64ee-11e8-9813-f172b7c4374d.png">

Reply:
Hmm, true, but it's still inaccurate as the element could be hidden by other means. (i.e. recursively hidden by parents, etc.)

Reply:
That's true. But I did not cover other hidden pattens (e.g. visibility:hidden, overlayed by other elements) because I did not see such examples in my analysis.
I left a TODO comment.